include/circt/Dialect/StaticLogic/StaticLogic.td

```diff
@@ -72,4 +72,159 @@ def ReturnOp : Op<StaticLogic_Dialect, "return", [Terminator]> {
+    both return results from the pipeline and register `iter_args`. Stages may
+    access SSA values dominating the pipeline, as well as `iter_args`, which are
+    block arguments.
+  }];
+
+  let arguments = (ins
+    Variadic<AnyType>:$iterArgs
+  );
+
+  let results = (outs
+    Variadic<AnyType>:$results
+  );
+
+  let regions = (region
+    SizedRegion<1>:$condition,
+    SizedRegion<1>:$stages
+  );
+
+  let parser = [{
+    return parse$cppClass(parser, result);
+  }];
+
+  let printer = [{
+    print$cppClass(p, *this);
+  }];
+
+  let verifier = [{
+    return ::verify$cppClass(*this);
+  }];
+}
+
+def PipelineStageOp : Op<StaticLogic_Dialect, "pipeline.stage",
+    [HasParent<"PipelineWhileOp">]> {
+  let summary = "StaticLogic dialect pipeline stage.";
+  let description = [{
+    This operation has a single-block region which dictates the operations that
+    may occur concurrently.
+
+    It may have an optional `when` predicate, in which case the stage only
+    executes when the predicate is `true : i1`, and pushes a bubble through the
+    pipeline otherwise.
+
+    It has a `staticlogic.register` terminator, which passes the concurrently
+    computed values forward to the next stage.
+
+    Any stage may access `iter_args`.
+
+    Other than `iter_args`, stages may only access SSA values dominating the
+    pipeline or SSA values computed by a previous stage. This ensures the stages
+    capture the coarse-grained schedule of the pipeline and how values feed
+    forward and backward.
+  }];
+
+  let arguments = (ins
+    Optional<I1>:$when
+  );
+
+  let results = (outs
+    Variadic<AnyType>:$results
+  );
+
+  let regions = (region
+    SizedRegion<1>:$body
+  );
+
+  let assemblyFormat = [{
+    (`when` $when^)? $body (`:` type($results)^)? attr-dict
+  }];
+}
+
+def PipelineRegisterOp : Op<StaticLogic_Dialect, "pipeline.register",
+    [ParentOneOf<["PipelineWhileOp", "PipelineStageOp"]>, Terminator]> {
+  let summary = "StaticLogic dialect pipeline register.";
+  let description = [{
+    The `staticlogic.pipeline.register` terminates a pipeline stage and
+    "registers" the values specified as operands. These values become the
+    results of the stage.
+  }];
+
+  let arguments = (ins
+    Variadic<AnyType>:$operands
+  );
+
+  let assemblyFormat = [{
+    $operands (`:` type($operands)^)? attr-dict
+  }];
+
+  let verifier = [{
+    return ::verify$cppClass(*this);
+  }];
+}
+
+def PipelineTerminatorOp : Op<StaticLogic_Dialect, "pipeline.terminator",
+    [HasParent<"PipelineWhileOp">, Terminator, AttrSizedOperandSegments]> {
+  let summary = "StaticLogic dialect pipeline terminator.";
+  let description = [{
+    The `staticlogic.pipeline.terminator` operation represents the terminator of
+    a `staticlogic.pipeline.while`.
+
+    The `results` section accepts a variadic list of values which become the
+    pipeline’s return values. These must be results of the last stage, and their
```

Comment:
The last-stage-constraint has been lifted, hasn't it?

Reply:
Hmm for `staticlogic.pipeline.terminator`, I was intending for results to always come from the final stage (I did miss a verifier for that). I lifted the constraint that stages must always access results from the stage before them, but for the outputs of the pipeline itself, shouldn't these always come from the last/final stage? I haven't really thought through what it would look like to output intermediate stages' results.

Reply:
> I did miss a verifier for that.

Ah ok, that led me to believe this was just an oversight in the docs.

Anyways, my thinking was, once a value is marked as registered, the lowering can preserve it until the end of the pipeline automatically. Allowing results to be computed in earlier stages could be useful if the pipeline returns multiple values that don't necessarily become available in the same stage, or if some stateful ops (e.g. a `memref.store`) occupy the last stage.

Reply:
Updated the wording to note the "final" stage for pipeline output results, and added a verifier to enforce this. If there is a use-case to allow intermediate stages' results to be output, we can walk this back, but I think this is a good place to start.

Reply:
Ah I had missed your previous comment @jopperm.

> Allowing results to be computed in earlier stages could be useful if the pipeline returns multiple values that don't necessarily become available in the same stage, or if some stateful ops (e.g. a memref.store) occupy the last stage.

I think this does make sense, as long as we document that the lowering must take this into account. I'll walk back my last change and update the docs.